fastapi_security/basic.py

```diff
@@ -36,3 +53,47 @@ def _make_credentials(
             c if isinstance(c, HTTPBasicCredentials) else HTTPBasicCredentials(**c)
             for c in credentials
         ]
+
+
+class BasicAuthWithDigestValidator:
+    def __init__(self):
+        self._salt = None
+        self._credentials = []
+
+    def init(self, salt: str, credentials: IterableOfHTTPBasicCredentialsDigest):
+        self._salt = salt
+        self._credentials = self._make_credentials(credentials)
+
+    def is_configured(self) -> bool:
+        return self._salt and len(self._credentials) > 0
+
+    def validate(self, credentials: HTTPBasicCredentials) -> bool:
+        if not self.is_configured():
+            return False
+        return any(
+            (
+                secrets.compare_digest(c.username, credentials.username)
+                and c.digest == self.generate_digest(self._salt, credentials.password)
+            )
+            for c in self._credentials
+        )
+
+    def generate_digest(self, secret: str):
+        if not self._salt:
+            raise ValueError('BasicAuthWithDigestValidator: cannot generate digest, salt is empty')
+        return generate_digest(self._salt, secret)
+
+    def _make_credentials(
+        self, credentials: IterableOfHTTPBasicCredentialsDigest
+    ) -> List[HTTPBasicCredentialsDigest]:
+        return [
+            c if isinstance(c, HTTPBasicCredentialsDigest) else HTTPBasicCredentialsDigest(**c)
+            for c in credentials
+        ]
+
+
+def generate_digest(salt: str, secret: str):
```

Comment:
Perhaps hashing method should be configurable?

Reply:
In general, yes.

But there are a few questions to be answered beforehand:

- what algorithms should this support? only built-ins or should `cryptography` package be covered, too?
- should this support algorithms that are known to be unsafe?
- should this support algorithms that don't accept salt?
- should this allow salt-less configuration for algorithms that can accept salt?

I would suggest going for maximum security right now, and wait for some feeback. If usage patterns indicate that more algorithms are required, it can always be added later on.